### PR TITLE
Add vmservice for android build options

### PR DESCRIPTION
**Pull request by chunhtai**
https://github.com/flutter/flutter/issues/120408

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/gradle/flutter.gradle
Comment (reidbaker) [inline packages/flutter_tools/gradle/flutter.gradle:974]: A bit more than a nitpick on the name. These are not buildOptions they are buildVariants. Why not create a gradle task that uses the existing verbiage? 

Similarly since this does not return anything the task should be printVariants or listVariants.

File: packages/flutter_tools/lib/src/android/android_builder.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/android_builder.dart:43]: I am not sure if on this function is the right place but please add some documentation for what "getBuildOptions" is for an who should use it.

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/gradle.dart:707]: Consider adding a comment `'-q', // suppresses gradle output.` for readability and because the rest of your code assumes no gradle logging.

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/gradle.dart:727]: I think you copied this from the aar task and if merged would pollute that data with information about the speed of this command. Please use a different label. 
I am not sure who looks at these usage numbers but they might have a suggestion for what to name this new value. 

DO NOT MERGE without changing this value.

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (chunhtai) [reply packages/flutter_tools/lib/src/android/gradle.dart:36]: I really dislike the fact I have to println in gradle and use regex like this. I feel there may be a better way, but I am not a gradle expert

File: packages/flutter_tools/lib/src/project.dart
Comment (chunhtai) [reply packages/flutter_tools/lib/src/project.dart:463]: I am also new to this @christopherfujino may know more. But this is what I assuming when I read the code.
AndroidProject class is responsible for all interactions from/to android sub project in the flutter folder. The vmservice is to register callback to the VMService so that whichever other dart running instances (in this case, Devtool) that connect to the VMService can call into the callbacks to get what it needs

File: packages/flutter_tools/test/general.shard/resident_web_runner_test.dart
Comment (chunhtai) [reply packages/flutter_tools/test/general.shard/resident_web_runner_test.dart:81]: This test assert all the vm services that will be registered when the vmservice.dart is run, since I added a new one, so i have to add a expect here.

**Comment by chunhtai** [comment]
A friendly bump

File: packages/flutter_tools/gradle/flutter.gradle
Comment (reidbaker) [reply packages/flutter_tools/gradle/flutter.gradle:979]: Also consider adding a comment that the the format is used in other code and should not change. 
If we were in google3 I would consider adding a Lint.IfChanged check with "packages/flutter_tools/lib/src/android/gradle.dart"

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/gradle.dart:736]: Is this required? If so why? We have a few on-going issues with how we get the version of java used to run android apps and I dont think we should introduce a new place where we make that determination. 

See this document https://docs.google.com/document/d/1hTXkjbUrBnXgu8NQsth1c3aEqo77rWoEj8CcsQ39wwQ/edit#

See this issue https://github.com/flutter/flutter/issues/122609 ,
this issue https://github.com/flutter/flutter/issues/121501 this feature request https://github.com/flutter/flutter/issues/106416 

As examples of why assuming the java version is dangerous

File: packages/flutter_tools/test/general.shard/resident_web_runner_test.dart
Comment (reidbaker) [reply packages/flutter_tools/test/general.shard/resident_web_runner_test.dart:81]: Thank you

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (reidbaker) [reply packages/flutter_tools/lib/src/android/gradle.dart]: I agree. For now let us document the expectations and move forward.

File: packages/flutter_tools/gradle/flutter.gradle
Comment (reidbaker) [reply packages/flutter_tools/gradle/flutter.gradle:979]: https://docs.gradle.org/current/userguide/more_about_tasks.html
"""
If you look at the API of the tasks container you may notice that there are additional methods to create tasks. The use of these methods is discouraged and will be deprecated in future versions. These methods only exist for backward compatibility as they were introduced before task configuration avoidance was added to Gradle.
"""
--


</div></div></div><div class="sect1" style="box-sizing: border-box; margin: 0px; padding: 0px; direction: ltr; border: 0px none;"><h2 id="sec:locating_tasks" style="box-sizing: border-box; margin: -1em 0px 0.5em; padding: 60px 0px 0px; direction: ltr; font-family: inherit; font-weight: 500; font-style: normal; color: inherit; text-rendering: optimizelegibility; line-height: 1.2; font-size: 1.5rem; word-spacing: -0.05em; letter-spacing: normal;"><a class="anchor" href="https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:locating_tasks" style="box-sizing: border-box; background: rgb(255, 255, 255); color: rgb(29, 162, 189); text-decoration: none; line-height: inherit; position: absolute; z-index: 1001; width: 1.5ex; margin-left: -1.5ex; display: inline-block; visibility: hidden; text-align: center; font-weight: normal; padding-top: 60px; margin-top: -60px; border: none; font-family: Lato, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-size: 24px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: -1.2px; -webkit-text-stroke-width: 0px;"></a><a class="link" href="https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:locating_tasks" style="box-sizing: border-box; background: rgb(255, 255, 255); color: rgb(2, 48, 58); text-decoration: none; line-height: inherit; font-family: Lato, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-size: 24px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 500; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: -1.2px; -webkit-text-stroke-width: 0px;"><br class="Apple-interchange-newline">If you look at the API of the tasks container you may notice that there are additional methods to create tasks. The use of these methods is discouraged and will be deprecated in future versions. These methods only exist for backward compatibility as they were introduced before [task configuration avoidance](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#task_configuration_avoidance) was added to Gradle.

Comment (chunhtai) [reply packages/flutter_tools/gradle/flutter.gradle:979]: I see, thanks for the link, I was wondering why there are some many way to create tasks, and not sure which one I should use...

Comment (chunhtai) [reply packages/flutter_tools/gradle/flutter.gradle:979]: So we don't want to include the task in lists, in that case how would the description be useful?

Comment (chunhtai) [reply packages/flutter_tools/gradle/flutter.gradle:979]: For some reason, this task is ran every time I ran `./gradlew tasks --all` in the android subproject, did i miss something?

File: packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart
Comment (reidbaker) [inline packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:814]: I don't think this or any other test validates the actual output of the task you wrote.

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (chunhtai) [reply packages/flutter_tools/lib/src/android/gradle.dart:736]: I am not entirely sure whether this is needed, I copied from the other `buildGradleAar`. Is there a prefer way to get the java version?

Comment (reidbaker) [reply packages/flutter_tools/lib/src/android/gradle.dart:736]: https://github.com/flutter/flutter/pull/123644 shows how to get the version of java gradle is using in gradle and exposes it as a task that dart code could rely on. 

here is the code that attempts to find the java binary that should be used. 
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/android/android_sdk.dart#L413

File: packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart
Comment (reidbaker) [reply packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:814]: See https://github.com/flutter/flutter/pull/123644 for an example integration test you can copy.

Comment (chunhtai) [reply packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:814]: done

File: packages/flutter_tools/gradle/flutter.gradle
Comment (gnprice) [inline packages/flutter_tools/gradle/flutter.gradle:697]: I believe this will print the output when the task is "configured", which may not mean it's actually going to get run. (See [Gradle docs on the distinction](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html).)

This might be related to why you were seeing the prints happen on `./gradlew tasks --all`: https://github.com/flutter/flutter/pull/123034#discussion_r1151228762
though now that you're using `tasks.register`, I would not expect that to happen. To reproduce the issue, though, it should suffice to [cause this task to be configured](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:task_configuration_avoidance_pitfalls) — for example, by adding
```groovy
printBuildVariants {
  // This block adds additional configuration to the "printBuildVariants" task.
  // (In this case, it's empty, so it adds no configuration.)
  // But it does so using the eager Gradle API, so as a side effect
  // it causes that task to be configured immediately.
}
```

To fix the issue, wrap the work of the task inside `doLast`, like in [this hello-world example](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:defining_tasks) from the Gradle docs. That'd look like:
```suggestion
        project.tasks.register("printBuildVariants") {
            description "Prints out all build variants for this Android project"
            doLast {
                project.android.applicationVariants.all { variant ->
                    println "BuildVariant: ${variant.name}";
                }
            }
```

File: packages/flutter_tools/lib/src/android/android_sdk.dart
Comment (gnprice) [inline packages/flutter_tools/lib/src/android/android_sdk.dart:486]: nit: perhaps put this just above `findJavaHome`. That also means `sdkManagerVersion` remains next to `sdkManagerEnv`.

File: packages/flutter_tools/lib/src/android/android_sdk.dart
Comment (gnprice) [inline packages/flutter_tools/lib/src/android/android_sdk.dart:412]: Factoring this out makes a lot of sense. It should get some dartdoc, though — particularly as its name sounds like it might just look at JAVA_HOME, but it actually does a couple more things.

```suggestion
  /// A value that would be appropriate to use as JAVA_HOME.
  ///
  /// This method considers, in order:
  /// * the JDK bundled with Android Studio, if one is found;
  /// * the JAVA_HOME in the ambient environment, if set;
  /// * on macOS, the value provided by the `java_home` command.
  static String? findJavaHome({
```

File: packages/flutter_tools/lib/src/android/android_sdk.dart
Comment (gnprice) [inline packages/flutter_tools/lib/src/android/android_sdk.dart:442]:
```suggestion
  /// The filesystem path of an appropriate `java` executable.
  ///
  /// This comes from [findJavaHome] if that method returns non-null,
  /// and otherwise from searching PATH.
  static String? findJavaBinary({
```

**Review by gnprice — COMMENTED** [review]
DBC

File: packages/flutter_tools/gradle/flutter.gradle
Comment (chunhtai) [reply packages/flutter_tools/gradle/flutter.gradle:697]: Thanks, this indeed fixed the issue.

**Comment by christopherfujino** [comment]
@andrewkolos can you take a look at this one?

**Comment by andrewkolos** [comment]
I went ahead and resolved a merge conflict this had with https://github.com/flutter/flutter/pull/124233.

FYI the newly added test (android_gradle_print_build_variants_test.dart) is failing.

**Comment by reidbaker** [comment]
FYI I am waiting on a ping to re-review this.

**Comment by chunhtai** [comment]
This is ready for another look

File: packages/flutter_tools/lib/src/android/android_sdk.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/android_sdk.dart:556]: Why this change from java binary to path?

File: packages/flutter_tools/lib/src/android/android_sdk.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/android_sdk.dart:494]: This log line and the comment no longer makes sense in the refactored method `findJavaHome`

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/gradle.dart:686]: Same here. At minimum use a constant in this file.

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/gradle.dart:738]: constant here.
If the `if` statement ensures that javaHome is not null then why are the bangs needed? Should you save globals.androidSdk.javaHome to a variable?

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/gradle.dart:755]: Does this work on windows or would the end character in the terminal be different?

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/gradle.dart:758]: is this ! dangerous?

File: packages/flutter_tools/lib/src/project.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/project.dart:478]: Consider adding a dart doc

File: packages/flutter_tools/lib/src/project.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/project.dart:479]: Can androidBuilder nullability be checked here so that the next line is androidBuilder.getBuildVariants.. without the ?? operator?

File: packages/flutter_tools/lib/src/vmservice.dart
Comment (reidbaker) [inline packages/flutter_tools/lib/src/vmservice.dart:327]: If `vmService.registerService('flutterGetAndroidBuildVariants'` and `vmService.registerServiceCallback('flutterGetAndroidBuildVariants',` need the same `flutterGetAndroidBuildVariants`. Consider using a constant.

File: packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart
Comment (reidbaker) [inline packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:790]: Should you add tests for any of the unhappy paths to make sure your error handling behaves consistently across refactors?

File: packages/flutter_tools/test/integration.shard/android_gradle_print_build_variants_test.dart
Comment (reidbaker) [inline packages/flutter_tools/test/integration.shard/android_gradle_print_build_variants_test.dart:55]: Consider using a raw string here. I think what you have is more readable but a raw string is easier to maintain.

**Review by reidbaker — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (chunhtai) [reply packages/flutter_tools/lib/src/android/gradle.dart:738]: I assume if we don't use this it would use the one in the system paths. Having this would prioritize the. AndroidStudio one?

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (chunhtai) [reply packages/flutter_tools/lib/src/android/gradle.dart:755]: good catch

File: packages/flutter_tools/lib/src/project.dart
Comment (reidbaker) [reply packages/flutter_tools/lib/src/project.dart:478]: I think this could still benefit from a dart doc.

**Review by reidbaker — APPROVED** [review]

**Comment by reidbaker** [comment]
Please still get a review from someone on the tools team to check the non android/gradle specific parts.

**Review by andrewkolos — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
change requires g3fix cl/525157717

**Pull request merged**
